xl: Delete older data-dir when replacing an existing version-id

vadmeste opened the pull request:
## Description

This commit adds a defensive code which removes an existing data-dir if
we are overwriting an existing version-id.

Though this use case is avoided by recent commits, this commit is only a
defensive code to avoid having stale data dirs (xl.meta not having a
reference to it)

## Motivation and Context
Add a defensive code to avoid having stale data dir

## How to test this PR?
Not trivial

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

harshavardhana commented on cmd/xl-storage.go:
```diff
@@ -2329,35 +2329,41 @@ func (s *xlStorage) RenameData(ctx context.Context, srcVolume, srcPath string, f
+		reqVID = fi.VersionID
+	}
+
+	ofi, err := xlMeta.ToFileInfo(dstVolume, dstPath, reqVID)
+	if err == nil && !ofi.Deleted {
+		if xlMeta.SharedDataDirCountStr(reqVID, ofi.DataDir) == 0 {
+			// Purge the destination path as we are not preserving anything
+			// versioned object was not requested.
+			oldDstDataPath = pathJoin(dstVolumeDir, dstPath, ofi.DataDir)
+			// if old destination path is same as new destination path
+			// there is nothing to purge, this is true in case of healing
+			// avoid setting oldDstDataPath at that point.
+			if oldDstDataPath == dstDataPath {
+				oldDstDataPath = ""
+			} else {
+				xlMeta.data.remove(reqVID, ofi.DataDir)
```
This data-dir fix can be sent separately - this is not dependent on `PutObject()` precondition check.

vadmeste replied:
> This data-dir fix can be sent separately - this is not dependent on PutObject() precondition check.

Well, disabling the precondition check is needed unless we remove the retry of minio-go in replication (like adding a retry option)

Currently, let's say if an object is replicated but the source server never receives a success message (networking issue).. then it will retry to replicate indefinitely without success. The reason of the indefinite retry is that replicating over an existing version in the target cluster will fail with Precondition error (because the precondition check says that source and target version are exactly the same)

harshavardhana replied:
Correct in such a scenario replication must mark that as "completed" @vadmeste PUT will never return a response as "precondition" check

harshavardhana replied:
> Well, disabling the precondition check is needed unless we remove the retry of minio-go in replication (like adding a retry option)


We have no retry enabled in minio-go for Replication 

> cmd/common-main.go:     minio.MaxRetry = 1

poornas replied:
I do not follow this PR - if incoming replication failed to respond successfully for whatever reason, the re-replication or healing attempt will go to replicateAll which will do a stat first and update the replication status to COMPLETED.  The pre-condition check in replicateObject applies only to incoming requests for the first attempt.

harshavardhana replied:
@poornas the main fix in this PR is xl-storage.go change I don't think other handler layer changes are needed

Commit:
xl: Delete older data-dir when replacing an existing version-id

This commit adds a defensive code which removes an existing data-dir if
we are overwriting an existing version-id.

Though this use case is avoided by recent commits, this commit is only a
defensive code to avoid having stale data dirs (xl.meta not having a
reference to it)
